### PR TITLE
fix(optimizer): incorrect merge with literal alias collision

### PR DESCRIPTION
**Pull request by fivetran-kwoodbeck**
Merging a derived table that projects a numeric literal (`6 AS a`) whose alias matches a `FROM` column, and is used in the outer `GROUP BY`, rewrites it to a bare `GROUP BY` a that rebinds to the column instead of the constant.  This changes the query result.  The fix is that `_mergeable` now blocks those merges.

Sample query:
```
SELECT s.a, SUM(s.b) AS b FROM (SELECT 6 AS a, b FROM x) AS s GROUP BY s.a;
```
Before, this query will return different results:
```
SELECT 6 AS a, SUM(x.b) AS b FROM x AS x GROUP BY a;
```
After:
```
SELECT s.a AS a, SUM(s.b) AS b FROM (SELECT 6 AS a, x.b AS b FROM x AS x) AS s GROUP BY s.a;
```

### CODE REVIEW

File: sqlglot/optimizer/merge_subqueries.py [hunk 1]
```diff
@@ -153,6 +153,32 @@ def _is_a_window_expression_in_unmergable_operation():
             for column in unmergable_window_columns
         )
 
+    def _literal_group_alias_collision():
+        """A numeric-literal projection in the outer GROUP BY merges to a bare `GROUP BY a`
+        that can collied with a same-named FROM column."""
```
Comment (georgesittas) [inline sqlglot/optimizer/merge_subqueries.py]:
```suggestion
        that can collide with a same-named FROM column."""
```

File: sqlglot/optimizer/merge_subqueries.py [hunk 2]
```diff
@@ -153,6 +153,32 @@ def _is_a_window_expression_in_unmergable_operation():
...
+        group = outer_scope.expression.args.get("group")
+        if not group:
+            return False
+        literal_aliases = {s.alias_or_name for s in inner_select.selects if s.unalias().is_number}
+        if not literal_aliases:
+            return False
```
Comment (georgesittas) [inline sqlglot/optimizer/merge_subqueries.py]: We should be careful not to walk `selects` / `expressions` more than needed. This can be wasteful if repeated without care. It's worth looking into refactoring this.

File: sqlglot/optimizer/merge_subqueries.py [hunk 3]
```diff
@@ -153,6 +153,32 @@ def _is_a_window_expression_in_unmergable_operation():
...
+        inner_name = from_or_join.alias_or_name
+        grouped = {
+            c.name
+            for c in group.find_all(exp.Column)
+            if c.table == inner_name and c.name in literal_aliases
+        }
```
Comment (georgesittas) [inline sqlglot/optimizer/merge_subqueries.py:188]: Similarly, these `find` / `find_all` calls can result in walking many nodes. There are other places where they happen in `_mergeable`. We should make sure we're not repeating a lot of work.

There may be a better alternative of pre-aggregating everything we need for these checks, but for that we would need to run all the cheap "mergeable" checks first, to avoid taking on the overhead preemptively.

File: sqlglot/optimizer/merge_subqueries.py [hunk 4]
```diff
@@ -153,6 +153,32 @@ def _is_a_window_expression_in_unmergable_operation():
...
+        if not grouped:
+            return False
+        for source in inner_scope.sources.values():
+            if isinstance(source, exp.Table):
+                return True
+            if isinstance(source, Scope) and grouped & {
+                s.alias_or_name for s in source.expression.selects
+            }:
```
Comment (georgesittas) [inline sqlglot/optimizer/merge_subqueries.py]: What about:
```suggestion
            if isinstance(source, Scope) and grouped & set(source.expression.named_selects):
```

**Review by georgesittas — COMMENTED** [review]
Tried optimizing this query as well

```sql
-- Input
WITH c AS (SELECT DISTINCT x.b AS b FROM x)
SELECT s.a, SUM(s.b) AS b FROM (SELECT 6 AS a, b FROM c) AS s CROSS JOIN x GROUP BY s.a

-- Optimized under this PR
WITH c AS (...) SELECT 6 AS a, SUM(c.b) AS b FROM c CROSS JOIN x GROUP BY a
```

The `GROUP BY` now binds `a` to `x.a` from the outer cross join. Seems like a similar bug. Additionally, there appears to be a similar bug with `ORDER BY` on a non-projected literal alias:

```sql
-- input
SELECT SUM(s.b) AS b FROM (SELECT 6 AS a, b FROM x) AS s GROUP BY s.a ORDER BY s.a

-- Optimized
SELECT SUM(x.b) AS b FROM x AS x GROUP BY a ORDER BY 6
```

This is also problematic, but let's address it separately to avoid scope creep.

File: sqlglot/optimizer/merge_subqueries.py (hunk 3 above)
Comment (fivetran-kwoodbeck) [reply sqlglot/optimizer/merge_subqueries.py:188]: found a way around the `find_all`

File: sqlglot/optimizer/merge_subqueries.py (hunk 2 above)
Comment (fivetran-kwoodbeck) [reply sqlglot/optimizer/merge_subqueries.py]: I'm not sure we can get around this one.

File: sqlglot/optimizer/merge_subqueries.py [hunk 5]
```diff
@@ -159,6 +159,36 @@ def _window_projection_blocks_merge():
...
+        literal_grouped = {
+            s.alias_or_name
+            for s in inner_select.selects
+            if s.alias_or_name in grouped and s.unalias().is_number
+        }
+        if not literal_grouped:
+            return False
+        # `GROUP BY a` resolves against the whole merged FROM: inner sources + outer's others.
+        merged_sources = [src for name, src in outer_scope.sources.items() if name != inner_name]
+        merged_sources.extend(inner_scope.sources.values())
```
Comment (georgesittas) [inline sqlglot/optimizer/merge_subqueries.py]: Shouldn't we iterate over `selected_sources` here? The bare `sources` contains CTEs from outer scopes that may not be selected, so we could get a non-empty intersection between `literal_grouped` and an irrelevant CTE, leading to an over-conservative merge block. Try this:

```sql
WITH unused AS (SELECT 1 AS a), t1 AS (SELECT 1 AS col)
SELECT t.a, SUM(t.b) AS b FROM (SELECT 6 AS a, t1.col AS b FROM t1) AS t GROUP BY t.a
```

Also, should we make `merged_sources` a set? We could end up walking sources >1 times if they're referenced in both the inner and the outer scope.

File: sqlglot/optimizer/merge_subqueries.py [hunk 6]
```diff
@@ -159,6 +159,36 @@ def _window_projection_blocks_merge():
...
+        for source in merged_sources:
+            if isinstance(source, exp.Table):
+                return True
```
Comment (georgesittas) [inline sqlglot/optimizer/merge_subqueries.py]: We could also just pass a schema to `merge_subqueries` to make this less conservative. So, get an optional schema (to make this PR backwards-compatible) and then if it's present, look up its columns.

File: sqlglot/optimizer/merge_subqueries.py [hunk 7]
```diff
@@ -159,6 +159,36 @@ def _window_projection_blocks_merge():
...
+        """A numeric-literal projection in the outer GROUP BY merges to a bare `GROUP BY a`
+        that can collide with a same-named FROM column."""
+        group = outer_scope.expression.args.get("group")
+        if not group:
+            return False
+        # Only a bare top-level `GROUP BY <col>` key can hit the literal rewrite.
+        inner_name = from_or_join.alias_or_name
+        grouped = {
+            e.name for e in group.expressions if isinstance(e, exp.Column) and e.table == inner_name
+        }
```
Comment (georgesittas) [inline sqlglot/optimizer/merge_subqueries.py:188]: What if there's a `Column` nested in a `ROLLUP(...)` or something along those lines? We'll miss any such references that aren't right beneath the `Group` node, right? Perhaps finding all columns under every grouped expression is the right way to go here.

Comment (fivetran-kwoodbeck) [reply sqlglot/optimizer/merge_subqueries.py:188]: I wasn't aware `ROLLUP` existed, I'll have it search for nested `Column`s.

File: sqlglot/optimizer/merge_subqueries.py [hunk 8]
```diff
@@ -159,6 +173,45 @@ def _window_projection_blocks_merge():
...
+        merged_sources = {
+            src for name, (_, src) in outer_scope.selected_sources.items() if name != inner_name
+        }
+        merged_sources.update(src for _, src in inner_scope.selected_sources.values())
+        for source in merged_sources:
+            if isinstance(source, exp.Table):
+                if schema is None:
+                    return True
+                if literal_grouped & set(schema.column_names(source)):
+                    return True
```
Comment (georgesittas) [inline sqlglot/optimizer/merge_subqueries.py]: This is incorrect in the case where either there's no schema or the schema is empty, because the intersection will produce a falsey value and we'll end up with the same problematic post-merge query.

The proper way to encode this condition is:

```python
for source in merged_sources:
    if isinstance(source, exp.Table):
        columns = schema.column_names(source) if schema else None
        if not columns or literal_grouped & set(columns):
            return True
    elif isinstance(source, Scope) and literal_grouped & set(source.expression.named_selects):
        return True
return False
```

We should make sure to cover these cases in the tests as well.

**Review by georgesittas — CHANGES_REQUESTED** [review]

File: sqlglot/optimizer/merge_subqueries.py (hunk 8 above)
Comment (fivetran-kwoodbeck) [reply sqlglot/optimizer/merge_subqueries.py]: We handle no schema with the `if schema is None` branch above.  I was assuming that the schema would be either absent or a full schema, so I'll update this and add tests for empty / partial schemas.

Comment (georgesittas) [reply sqlglot/optimizer/merge_subqueries.py]: The schema is expected to be supplied & complete, that's true. But in this case it's a small condition rewrite, so worth doing it imo.

**Comment by georgesittas** [comment]
Feel free to merge after resolving the schema comment and rebasing. Nice work!
